# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
os.getcwd()

'C:\\Users\\User\\Desktop\\Guided Capstone Project'

In [3]:
os.listdir()

['.ipynb_checkpoints',
 'data',
 'figures',
 'GuidedCapstoneStep3HL.ipynb',
 'GuidedCapstoneStep4 and step5HL.ipynb',
 'GuidedCapstone_Step2HL (1).ipynb',
 'models',
 'updated_ski_data.csv']

In [4]:
path = 'C:\\Users\\User\\Desktop\\Guided Capstone Project\\data'
os.chdir(path)
os.getcwd()

'C:\\Users\\User\\Desktop\\Guided Capstone Project\\data'

In [5]:
os.listdir()

['step2_output.csv', 'step3_output.csv']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [6]:
dfno = pd.read_csv('step3_output.csv')
dfno.head()

Unnamed: 0                           Name       state  summit_elev  \
0           0               Hilltop Ski Area      Alaska         2090   
1           1            Sunrise Park Resort     Arizona        11100   
2           2  Yosemite Ski & Snowboard Area  California         7800   
3           3         Boreal Mountain Resort  California         7700   
4           4                    Dodge Ridge  California         8200   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0            294      0        0.0          0          0     0  ...   
1           1800      0        0.0          0          1     2  ...   
2            600      0        0.0          0          0     0  ...   
3            500      0        0.0          0          1     1  ...   
4           1600      0        0.0          0          0     1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 27 columns]

In [7]:
dfno = dfno.drop('Unnamed: 0',axis=1)

In [8]:
dfno.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [9]:
dfo=dfno['state']
dfo

0          Alaska
1         Arizona
2      California
3      California
4      California
          ...    
171       Wyoming
172       Wyoming
173       Wyoming
174       Wyoming
175       Wyoming
Name: state, Length: 176, dtype: object

In [10]:
df_dum = pd.get_dummies(dfo)
df_dum.head()

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0       1        0           0         0            0      0         0   
1       0        1           0         0            0      0         0   
2       0        0           1         0            0      0         0   
3       0        0           1         0            0      0         0   
4       0        0           1         0            0      0         0   

   Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0        0     0      0  ...             0             0          0     0   
1        0     0      0  ...             0             0          0     0   
2        0     0      0  ...             0             0          0     0   
3        0     0      0  ...             0             0          0     0   
4        0     0      0  ...             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0  

[5 rows x 35 columns]

In [11]:
dfno = pd.concat([dfno.drop('state',axis=1),df_dum],axis=1)

In [12]:
dfno.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [13]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = dfno.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = dfno.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [14]:
X_scaled

array([[-0.49603034, -0.94696389,  0.        , ..., -0.13168538,
        -0.25819889, -0.17099639],
       [ 2.29154853,  1.6034447 ,  0.        , ..., -0.13168538,
        -0.25819889, -0.17099639],
       [ 1.2705707 , -0.42875338,  0.        , ..., -0.13168538,
        -0.25819889, -0.17099639],
       ...,
       [ 1.2730458 ,  1.21563357,  0.        , ..., -0.13168538,
        -0.25819889,  5.84807661],
       [ 1.8469591 ,  0.23171099,  0.        , ..., -0.13168538,
        -0.25819889,  5.84807661],
       [ 1.79652898,  0.41799582,  0.        , ..., -0.13168538,
        -0.25819889,  5.84807661]])

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [15]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [16]:
print(y)

[34.         78.         47.         64.16681004 78.         75.
 64.16681004 59.         84.         83.         64.16681004 79.
 50.         69.         89.         71.         56.         84.
 65.         65.         60.         55.         62.         42.
 42.         47.         32.         47.         64.16681004 43.
 55.         35.         40.         45.         52.         39.
 40.         46.         43.         55.         49.         32.
 75.         79.         68.         45.         40.         64.16681004
 47.         35.         65.         49.         49.         64.16681004
 49.         34.         45.         45.         45.         35.
 42.         45.         64.16681004 45.         47.         44.
 37.         58.         35.34       48.         59.         49.
 47.         48.         39.         50.         47.         70.
 89.         59.         64.16681004 50.         92.         58.
 93.         72.         93.         45.         89.         64.16681004
 

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [17]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [18]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [19]:
y_pred

array([ 5.06228294e+01,  5.82395287e+01,  4.70530052e+01,  4.47523460e+01,
        5.88567162e+01,  4.75207787e+01,  5.52729759e+01,  8.31784477e+01,
        3.69176292e+01,  4.53288841e+01,  7.89508324e+01,  8.23692894e+01,
        4.71198997e+01,  6.02224388e+01,  6.00442162e+01,  4.25075951e+01,
        4.21350365e+01,  3.26814232e+01,  3.43396263e+01,  3.76682396e+01,
        4.80862083e+01,  5.08998833e+01,  5.85847435e+01,  5.53042259e+01,
        3.27822533e+01,  7.09861107e+01,  6.00613060e+01,  3.08232689e+01,
        5.83114281e+01,  6.21252708e+01,  4.77966576e+01,  8.18334007e+01,
        8.97678032e+01,  7.03918724e+01,  5.29456596e+01,  7.74924583e+01,
       -3.52981007e+12,  7.87063255e+01,  6.17961693e+01,  4.90237083e+01,
        5.31828880e+01,  3.45787621e+01,  6.64704857e+01,  3.23759575e+01])

In [29]:
y_test

array([47.        , 62.        , 65.        , 48.        , 64.16681004,
       45.        , 59.        , 89.        , 49.        , 54.        ,
       79.        , 78.        , 40.        , 75.        , 68.        ,
       43.        , 75.        , 39.        , 39.        , 35.        ,
       49.        , 50.        , 60.        , 48.        , 49.        ,
       72.        , 64.16681004, 43.        , 56.        , 59.        ,
       47.        , 72.        , 79.        , 75.        , 59.        ,
       77.        , 64.16681004, 73.        , 49.        , 47.        ,
       49.        , 35.        , 63.2       , 55.        ])

## Review Model Outcomes — Iterate over additional models as needed

In [20]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [22]:
y_score = model.score(X_train,y_train)

In [23]:
y_score

0.9137556569263675

In [24]:
y_v_1= explained_variance_score(y_test, y_pred)
y_v_1

-1.4876995714877353e+21

In [30]:
y_m_1 = mean_absolute_error(y_test, y_pred)
y_m_1

80222956105.49896

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [27]:
lm_intercept = model.intercept_

In [28]:
lm_intercept

-20055738968.495335

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [31]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(model.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        3.961375e-01
vertical_drop      3.077587e+00
trams              3.648050e+11
fastEight          6.524533e+12
fastSixes          7.642412e+12
fastQuads          6.653224e+11
quad               7.960293e+11
triple             1.398003e+12
double             1.711529e+12
surface            1.637901e+12
total_chairs       3.117620e+12
Runs               4.519348e-01
TerrainParks       1.797363e+00
LongestRun_mi      1.089111e+00
SkiableTerrain_ac  1.676514e+00
Snow Making_ac     6.665039e-01
daysOpenLastYear   1.140747e+00
yearsOpen          3.879395e-01
averageSnowfall    1.895752e+00
AdultWeekday       9.336670e+00
projectedDaysOpen  2.112305e+00
NightSkiing_ac     9.194336e-01
clusters           1.561646e+00
Alaska             2.653125e+11
Arizona            2.653125e+11
California         8.897157e+11
Colorado           5.864375e+11
Connecticut        5.864375e+11
Idaho              6.405290e+11
Illinois           4.569013e+11
Indiana            3.741350e+11
Iowa               4.569013e+11
Maine              5.864375e+11
Maryland           2.653125e+11
Massachusetts      5.260571e+11
Michigan           9.855922e+11
Minnesota          7.352556e+11
Missouri           3.741350e+11
Montana            5.260571e+11
Nevada             2.653125e+11
New Hampshire      8.544310e+11
New Jersey         1.024099e+04
New Mexico         6.405290e+11
New York           1.120253e+12
North Carolina     4.569013e+11
Ohio               5.260571e+11
Oregon             4.569013e+11
Pennsylvania       8.544310e+11
Rhode Island       2.653125e+11
South Dakota       2.653125e+11
Tennessee          2.653125e+11
Utah               4.569013e+11
Vermont            7.352556e+11
Virginia           3.741350e+11
Washington         5.260571e+11
West Virginia      4.569013e+11
Wisconsin          8.544310e+11
Wyoming            5.864375e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [32]:
dfno.head(10)

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   
5               Donner Ski Ranch         8012            750      0   
6                  June Mountain        10090           2590      0   
7            Mt. Shasta Ski Park         6890           1435      0   
8                  Mountain High         8200           1600      0   
9                 Ski China Peak         8709           1679      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   
5        0.0          0          0     0       1       5  ...             0   
6        0.0          0          2     0       0       4  ...             0   
7        0.0          0          0     0       3       0  ...             0   
8        0.0          0          2     2       2       5  ...             0   
9        0.0          0          0     1       4       2  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   
5             0          0     0        0         0           0   
6             0          0     0        0         0           0   
7             0          0     0        0         0           0   
8             0          0     0        0         0           0   
9             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  
5              0          0        0  
6              0          0        0  
7              0          0        0  
8              0          0        0  
9              0          0        0  

[10 rows x 60 columns]

In [33]:
#here dropping dummy state variable from dataset 'dfno'
dfno = dfno.drop(df_dum,axis=1)

In [34]:
dfno.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       1       0  ...   
1        0.0          0          1     2       3       1  ...   
2        0.0          0          0     0       1       3  ...   
3        0.0          0          1     1       3       1  ...   
4        0.0          0          0     1       2       5  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 25 columns]

In [36]:
X2 = dfno.drop(['Name','AdultWeekend'], axis=1)
y2 = dfno.AdultWeekend
scaler2 = StandardScaler().fit(X2)
X2_scaled=scaler2.transform(X2) 

In [37]:
X2_scaled

array([[-0.49603034, -0.94696389,  0.        , ...,  1.9013763 ,
        -0.30586211, -0.76462969],
       [ 2.29154853,  1.6034447 ,  0.        , ..., -0.39112723,
         0.75632675,  0.58111856],
       [ 1.2705707 , -0.42875338,  0.        , ..., -0.24784576,
        -0.94317542,  0.58111856],
       ...,
       [ 1.2730458 ,  1.21563357,  0.        , ...,  0.51632208,
         1.39364006,  0.58111856],
       [ 1.8469591 ,  0.23171099,  0.        , ...,  0.37557209,
        -0.94317542,  0.58111856],
       [ 1.79652898,  0.41799582,  0.        , ...,  0.37557209,
        -0.94317542,  0.58111856]])

In [38]:
#splitting data into trains and tests
y2 = y2.ravel()
X_train, X_test, y_train, y_test = train_test_split(X2_scaled, y2, test_size=0.25, random_state=1)

In [44]:
# model-2
lm = linear_model.LinearRegression()
model2 = lm.fit(X_train,y_train)

In [45]:
# Predicting the valyes of y using X_test
y_pred2 = model2.predict(X_test)
y_pred2

array([53.90401302, 56.24971539, 40.1601437 , 50.90139668, 60.88786166,
       47.87237028, 52.34516943, 74.90701923, 37.50895989, 45.64447885,
       75.59328597, 78.04980122, 47.14896108, 62.56332424, 59.45968503,
       48.61841788, 46.73699381, 41.17902216, 46.12098935, 40.86261933,
       48.44508742, 54.23245195, 57.71515226, 61.07076437, 35.64947554,
       62.59536216, 53.93399851, 37.27614857, 50.40500693, 58.88561327,
       53.19268748, 74.74227585, 95.46457734, 67.71100976, 56.44576039,
       78.39876585, 52.72642025, 87.66912359, 54.98441022, 50.17271473,
       52.01560946, 34.99762348, 62.39082499, 38.61261325])

In [46]:
y_score2 = model2.score(X_train,y_train)
y_score2

0.8363244732470392

In [47]:
y_v_2= explained_variance_score(y_test, y_pred2)
y_v_2

0.5325426531233332

In [48]:
y_m_2 = mean_absolute_error(y_test, y_pred2)
y_m_2

7.37396972614105

In [75]:
intercept2 = model.intercept_
intercept2

56.35202406990897

In [52]:
pd.DataFrame(abs(model2.coef_), X2.columns, columns=['Coefficient'])

Coefficient
summit_elev        2.988854e+00
vertical_drop      2.687827e+00
trams              0.000000e+00
fastEight          1.776357e-15
fastSixes          8.881784e-16
fastQuads          6.012949e-01
quad               2.053486e+00
triple             5.133334e-01
double             5.179764e-01
surface            1.312435e+00
total_chairs       4.776799e-01
Runs               5.624883e-02
TerrainParks       2.558779e+00
LongestRun_mi      1.272038e+00
SkiableTerrain_ac  1.519416e+00
Snow Making_ac     7.612125e-01
daysOpenLastYear   1.820064e+00
yearsOpen          3.913569e-01
averageSnowfall    7.430559e-01
AdultWeekday       1.015013e+01
projectedDaysOpen  1.692985e+00
NightSkiing_ac     2.114992e-01
clusters           3.521296e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [56]:
dfno_model3 = dfno.drop('summit_elev',axis=1)

In [61]:
dfno_model3.head()

Name  vertical_drop  trams  fastEight  fastSixes  \
0               Hilltop Ski Area            294      0        0.0          0   
1            Sunrise Park Resort           1800      0        0.0          0   
2  Yosemite Ski & Snowboard Area            600      0        0.0          0   
3         Boreal Mountain Resort            500      0        0.0          0   
4                    Dodge Ridge           1600      0        0.0          0   

   fastQuads  quad  triple  double  surface  ...  SkiableTerrain_ac  \
0          0     0       1       0        2  ...               30.0   
1          1     2       3       1        0  ...              800.0   
2          0     0       1       3        1  ...               88.0   
3          1     1       3       1        2  ...              380.0   
4          0     1       2       5        4  ...              862.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0       30.000000        150.000000       36.0             69.0          30.0   
1       80.000000        115.000000       49.0            250.0          74.0   
2      174.873239        110.000000       84.0            300.0          47.0   
3      200.000000        150.000000       54.0            400.0          49.0   
4      174.873239        115.103943       69.0            350.0          78.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0      34.00000              152.0            30.0         0  
1      78.00000              104.0            80.0         1  
2      47.00000              107.0             0.0         1  
3      64.16681              150.0           200.0         1  
4      78.00000              140.0             0.0         1  

[5 rows x 24 columns]

In [62]:
X3 = dfno_model3.drop(['Name','AdultWeekend'], axis=1)
y3 = dfno.AdultWeekend
scaler3 = StandardScaler().fit(X3)
X3_scaled=scaler3.transform(X3) 

In [63]:
X3_scaled

array([[-0.94696389,  0.        ,  0.        , ...,  1.9013763 ,
        -0.30586211, -0.76462969],
       [ 1.6034447 ,  0.        ,  0.        , ..., -0.39112723,
         0.75632675,  0.58111856],
       [-0.42875338,  0.        ,  0.        , ..., -0.24784576,
        -0.94317542,  0.58111856],
       ...,
       [ 1.21563357,  0.        ,  0.        , ...,  0.51632208,
         1.39364006,  0.58111856],
       [ 0.23171099,  0.        ,  0.        , ...,  0.37557209,
        -0.94317542,  0.58111856],
       [ 0.41799582,  0.        ,  0.        , ...,  0.37557209,
        -0.94317542,  0.58111856]])

In [64]:
#splitting data into trains and tests
y3 = y3.ravel()
X_train, X_test, y_train, y_test = train_test_split(X3_scaled, y3, test_size=0.25, random_state=1)

In [67]:
# model-3
lm = linear_model.LinearRegression()
model3 = lm.fit(X_train,y_train)

In [68]:
# Predicting the valyes of y using X_test
y_pred3 = model3.predict(X_test)
y_pred3

array([57.52507803, 56.83913849, 39.12375616, 51.14741314, 58.0466969 ,
       48.62319073, 52.40742923, 77.25195338, 38.05273498, 45.42424746,
       78.81239996, 78.07408894, 47.66642525, 60.36720392, 58.82001215,
       47.66979013, 45.37030459, 40.76370609, 45.4162971 , 41.35576471,
       46.25656931, 53.92196592, 53.79598152, 58.42516068, 32.52780244,
       62.35019464, 52.89611994, 36.34132534, 54.74258132, 56.82906437,
       52.1848831 , 74.2195321 , 89.92813498, 68.23020691, 49.85780957,
       82.8605179 , 53.08190017, 86.29583475, 54.07151078, 50.21176476,
       49.29258231, 34.15028812, 61.42494096, 37.95924483])

In [71]:
y_score3 = model3.score(X_train,y_train)
y_score3

0.8245402325089572

In [73]:
y_v_3 = explained_variance_score(y_test,y_pred3)
y_v_3

0.5266658380003132

In [74]:
y_m_3 = mean_absolute_error(y_test,y_pred3)
y_m_3

7.645321928693992

In [76]:
intercept3 = model.intercept_
intercept3

56.35202406990897

In [78]:
pd.DataFrame(abs(model3.coef_), X3.columns, columns=['Coefficient'])

Coefficient
vertical_drop      2.652931e+00
trams              3.108624e-15
fastEight          2.886580e-15
fastSixes          2.664535e-15
fastQuads          7.917712e-01
quad               2.003488e+00
triple             3.262767e-01
double             6.143749e-01
surface            1.026970e+00
total_chairs       6.245785e-01
Runs               2.159080e-01
TerrainParks       2.567094e+00
LongestRun_mi      1.534258e+00
SkiableTerrain_ac  2.713687e+00
Snow Making_ac     1.130926e+00
daysOpenLastYear   1.434676e+00
yearsOpen          4.865562e-01
averageSnowfall    1.511774e+00
AdultWeekday       1.011740e+01
projectedDaysOpen  1.418353e+00
NightSkiing_ac     3.713621e-01
clusters           2.896518e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.4876995714877353e+21 |80222956105.49896  |-|
| Model 2. | 0.5325426531233332|56.35202406990897 |'state'|
| Model 3. | 0.5266658380003132|7.645321928693992 |'state','summit_elev','base_elev'|

# Model Selection:

##Model 3 is selected